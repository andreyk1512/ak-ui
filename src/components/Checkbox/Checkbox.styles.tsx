import { css } from '@emotion/react';
import CheckedSVG from '../../assets/icons/checked.svg';
import * as styles from '../../styles';
import { CheckboxProps } from './Checkbox.types';

export const labelStyles = ({ disabled }: Pick<CheckboxProps, 'disabled'>) => css`
  font: inherit;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  color: ${styles.color.slateGray};
  cursor: pointer;
  transition: all 0.2s;

  ${disabled &&
  css`
    opacity: 0.65;
    cursor: not-allowed;
  `}
`;

export const inputStyles = css`
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  -webkit-appearance: none;
  appearance: none;
  background-color: ${styles.color.white};
  margin: 0 10px 0 0;
  width: 19px;
  height: 19px;
  border: 1px solid ${styles.color.beauBlue};
  border-radius: 3px;
  box-sizing: border-box;
  transition: transform linear 0.2s;

  &:active {
    transform: scale(0.98);
  }

  &:checked {
    border: none;
    background-color: ${styles.color.lightSeaGreen};

    &:hover {
      background-color: ${styles.color.greenCrayola};
    }

    &:before {
      position: absolute;
      content: '';
      width: 11px;
      height: 8px;
      display: inline-block;
      line-height: 100%;
      background: url(${CheckedSVG.toString()});
    }
  }

  &:checked + span {
    color: ${styles.color.yankeesBlue};
  }
`;
