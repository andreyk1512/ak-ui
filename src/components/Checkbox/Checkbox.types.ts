import { ChangeEventHandler, ReactNode } from 'react';

export type CheckboxProps = {
  name?: string;
  label: ReactNode;
  value?: boolean;
  disabled?: boolean;
  onChange?: ChangeEventHandler<HTMLInputElement>;
};
