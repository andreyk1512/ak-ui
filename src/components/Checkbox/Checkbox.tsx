import { useId } from 'react';
import { inputStyles, labelStyles } from './Checkbox.styles';
import { CheckboxProps } from './Checkbox.types';

export const Checkbox = ({ name, label, value, disabled, onChange }: CheckboxProps) => {
  const id = useId();

  return (
    <label htmlFor={id} css={labelStyles({ disabled })}>
      <input
        id={id}
        name={name}
        type="checkbox"
        checked={value}
        disabled={disabled}
        onChange={onChange}
        css={inputStyles}
      />
      <span>{label}</span>
    </label>
  );
};
