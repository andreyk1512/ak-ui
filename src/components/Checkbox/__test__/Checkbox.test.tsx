import { fireEvent, render } from '@testing-library/react';
import { Checkbox } from '../Checkbox';

describe('Checkbox', () => {
  it('should correctly render checkbox element with content', () => {
    const label = 'CheckBox';
    const { container } = render(<Checkbox label={label} />);

    expect(container.querySelector('label>span')?.textContent).toBe(label);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able to change checked value', () => {
    const ON_CHANGE = jest.fn();
    const { container, rerender } = render(<Checkbox label="CheckBox" onChange={ON_CHANGE} value={false} />);

    expect(container.querySelector('input')?.checked).toEqual(false);

    rerender(<Checkbox label="CheckBox" onChange={ON_CHANGE} value={true} />);

    expect(container.querySelector('input')?.checked).toEqual(true);
  });

  it('should be able to render a disabled checkbox', () => {
    const ON_CHANGE = jest.fn();
    const { container } = render(<Checkbox label="CheckBox" disabled onChange={ON_CHANGE} />);

    fireEvent.click(container.querySelector('label')!);

    expect(ON_CHANGE).not.toHaveBeenCalled();
  });
});
