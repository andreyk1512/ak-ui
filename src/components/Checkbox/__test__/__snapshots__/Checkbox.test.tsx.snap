// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render the correct snapshot of the checked state 1`] = `
<label
  class="css-1vof294-Checkbox"
  for=":r0:"
>
  <input
    checked=""
    class="css-l7v0wq-Checkbox"
    id=":r0:"
    type="checkbox"
  />
  <span>
    My Checkbox
  </span>
</label>
`;

exports[`Checkbox should render the correct snapshot of the disabled state 1`] = `
<label
  class="css-1okvwha-Checkbox"
  for=":r0:"
>
  <input
    class="css-l7v0wq-Checkbox"
    disabled=""
    id=":r0:"
    type="checkbox"
  />
  <span>
    My Checkbox
  </span>
</label>
`;

exports[`Checkbox should render the correct snapshot of the unchecked state 1`] = `
<label
  class="css-1vof294-Checkbox"
  for=":r0:"
>
  <input
    class="css-l7v0wq-Checkbox"
    id=":r0:"
    type="checkbox"
  />
  <span>
    My Checkbox
  </span>
</label>
`;
